### PR TITLE
C bindings - properties i/f

### PR DESCRIPTION
**Pull request by whatdoineed2do**
expose `properties` via `C` bindings interface

### CODE REVIEW

File: bindings/c/tag_c.cpp [hunk 1]
```diff
@@ -313,3 +314,162 @@ void taglib_id3v2_set_default_text_encoding(TagLib_ID3v2_Encoding encoding)
 
   ID3v2::FrameFactory::instance()->setDefaultTextEncoding(type);
 }
+
+
+/******************************************************************************
+ * Properties API
+ ******************************************************************************/
+namespace {
+PropertyMap  _taglib_property_get(TagLib_File *cfile)
```
Comment (ufleisch) [inline bindings/c/tag_c.cpp:323]: Do not use two spaces if one is enough.

File: bindings/c/tag_c.cpp [hunk 2]
```diff
@@ -313,3 +314,162 @@ void taglib_id3v2_set_default_text_encoding(TagLib_ID3v2_Encoding encoding)
...
+  const PropertyMap  map = _taglib_property_get(file);
+
+  TagLib::PropertyMap::ConstIterator  property = map.find(prop);
+  if (property == map.end())
+    return NULL;
+
+  char** props = (char**)malloc(sizeof(char*) * (property->second.size()+1) );
+  char** pp = props;
+
+  for (StringList::ConstIterator i= property->second.begin(); i!=property->second.end(); ++i) {
```
Comment (ufleisch) [inline bindings/c/tag_c.cpp:455]: Fix the spaces, it should be `for (StringList::ConstIterator i = property->second.begin(); i != property->second.end(); ++i) {`.

File: bindings/c/tag_c.cpp [hunk 3]
```diff
@@ -313,3 +314,162 @@ void taglib_id3v2_set_default_text_encoding(TagLib_ID3v2_Encoding encoding)
...
+    (*pp) = strdup(i->toCString());
```
Comment (ufleisch) [inline bindings/c/tag_c.cpp:456]: No need for these parentheses.

File: bindings/c/tag_c.h [hunk 4]
```diff
@@ -292,6 +292,27 @@ typedef enum {
 
 TAGLIB_C_EXPORT void taglib_id3v2_set_default_text_encoding(TagLib_ID3v2_Encoding encoding);
 
+/******************************************************************************
+ * Properties API
+ ******************************************************************************/
+
+/*!
+ * Sets the property with value.  value=null to remove, append to existing property or to replace
```
Comment (ufleisch) [inline bindings/c/tag_c.h:300]: Please try to have not more than 100 characters per line, I would even prefer 80. And use Doxygen `\return` and `\a` as in the other API comments.

File: bindings/c/tag_c.cpp [hunk 5]
```diff
@@ -422,15 +422,15 @@ char** taglib_property_keys(TagLib_File *file)
...
+  const PropertyMap  map = _taglib_property_get(file);
   if (map.isEmpty()) 
     return NULL;
 
   char** props = (char**)malloc(sizeof(char*)*(sizeof(char*) * (map.size()+1)) );
   char** pp = props;
 
-  for (const auto& i : map) {
-    (*pp) = strdup(i.first.toCString());
+  for (PropertyMap::ConstIterator i=map.begin(); i!=map.end(); ++i) {
```
Comment (ufleisch) [inline bindings/c/tag_c.cpp:432]: For TagLib 2.0, we now use C++17, so feel free to use the new features.

**Review by ufleisch — CHANGES_REQUESTED** [review]
Thanks for your contribution, this will certainly be a useful addition for the C bindings. Could you please rebase your branch to the current master and then fix it according to my review comments? For the style "violations", I only commented the first occurrence, but it would be nice if you could fix all occurrences.

**Comment by whatdoineed2do** [comment]
@ufleisch  - appreciate the feedback.  Rebased with master and style  items should be addressed now.

One issue I notice with the new `taglib_property_set*()` is that using `file->tag()->setProperties(map)` still does not yeild the desired result (non-standard map values disappear) as was with Taglib 1.x.  I dont know if this is pending implementation but i note that `Tag::setProperties()` is already `virtual`

File: bindings/c/tag_c.cpp [hunk 6]
```diff
@@ -313,3 +314,162 @@ void taglib_id3v2_set_default_text_encoding(TagLib_ID3v2_Encoding encoding)
...
+  if(dynamic_cast<const TrueAudio::File*>(file))
+    return dynamic_cast<const TrueAudio::File *>(file)->properties();
+  if(dynamic_cast<const Ogg::Vorbis::File* >(file))
+    return  dynamic_cast<const Ogg::Vorbis::File* >(file)->properties();
+  if(dynamic_cast<const WavPack::File* >(file))
+     return dynamic_cast<const WavPack::File* >(file)->properties();
+  if(dynamic_cast<const MP4::File* >(file))
+    return  dynamic_cast<const MP4::File* >(file)->properties();
+  if(dynamic_cast<const ASF::File* >(file))
+    return  dynamic_cast<const ASF::File* >(file)->properties();
```
[the remark this answers is not in the document]
Comment (whatdoineed2do) [reply bindings/c/tag_c.cpp]: dropped and using direct `file->tag()->properties()`

**Comment by whatdoineed2do** [comment]
When testing with mp3, mp4a and flac, all work as expected EXCEPT m4a.

I modified `tagreader_c.c` to test the set operation
```
diff --git a/examples/tagreader_c.c b/examples/tagreader_c.c
index 9adaf408..f939571e 100644
--- a/examples/tagreader_c.c
+++ b/examples/tagreader_c.c
@@ -97,6 +97,34 @@ int main(int argc, char *argv[])
       }
     }
 
+    {
+      printf("-- SET PROPERITES --\n");
+      const char* key = "taglibreader_c";
+      taglib_property_set(file, key, "new value");
+      taglib_property_set_append(file, key, "another value");
+
+       printf("== added 2x values\n");
+       propertyValues = pp1 = taglib_property_get(file, key);
+       while (pp1 && *pp1) {
+         printf("  \"%s\"\n", *pp1);
+         ++pp1;
+       }
+       taglib_property_free(propertyValues);
+
+      taglib_property_set(file, key, NULL);
+       propertyValues = taglib_property_get(file, key);
+       printf("== removed value, should be null - %x\n", propertyValues);
+
+       printf("== appending to empty\n");
+      taglib_property_set_append(file, key, "append value");
+       propertyValues = pp1 = taglib_property_get(file, key);
+       while (pp1 && *pp1) {
+         printf("  \"%s\"\n", *pp1);
+         ++pp1;
+       }
+       taglib_property_free(propertyValues);
+    }
+
     taglib_property_free(propertiesMap);
     taglib_tag_free_strings();
     taglib_file_free(file);
```


and the output:
```
******************** "/home/ray/dev/audiotag/tests/test.mp3" ********************
-- TAG --
title   - ""
artist  - "foo"
album   - ""
year    - "0"
comment - ""
track   - "0"
genre   - ""
-- AUDIO --
bitrate     - 129
sample rate - 44100
channels    - 2
length      - 0:03
-- PROPERITES --
"ARTIST"
  "foo"
-- SET PROPERITES --
== added 2x values
  "new value"
  "another value"
== removed value, should be null - 0
== appending to empty
  "append value"
******************** "/home/ray/dev/audiotag/tests/test.m4a" ********************
-- TAG --
title   - "MP4 ¯＼(º_o)/  file"
artist  - "foo"
album   - "audiotag test files"
year    - "2015"
comment - ""
track   - "0"
genre   - "Pop"
-- AUDIO --
bitrate     - 83
sample rate - 48000
channels    - 2
length      - 0:05
-- PROPERITES --
"ALBUM"
  "audiotag test files"
"ARTIST"
  "foo"
"DATE"
  "2015-10-21"
"ENCODEDBY"
  "Lavf58.76.100"
"GENRE"
  "Pop"
"TITLE"
  "MP4 �<(�_o)/� file"
-- SET PROPERITES --
== added 2x values
== removed value, should be null - 0
== appending to empty
******************** "/home/ray/dev/audiotag/tests/test.flac" ********************
-- TAG --
title   - "flac ¯\_(ツ)_/¯ file"
artist  - "foo"
album   - "audiotag test files"
year    - "1955"
comment - ""
track   - "4"
genre   - "Pop"
-- AUDIO --
bitrate     - 0
sample rate - 48000
channels    - 2
length      - 0:00
-- PROPERITES --
"ALBUM"
  "audiotag test files"
"ARTIST"
  "foo"
"DATE"
  "1955-11-05"
"ENCODER"
  "Lavf57.71.100"
"GENRE"
  "Pop"
"TITLE"
  "flac �\_(�)_/� file"
"TRACKNUMBER"
  "4/19"
-- SET PROPERITES --
== added 2x values
  "new value"
  "another value"
== removed value, should be null - 0
== appending to empty
  "append value"
```

**Pull request merged**